### ISSUE TITLE
SAML invalid destination caused by case sensitive ?

### ISSUE THREAD

**Issue by rvanderboom** [keycloak/keycloak#35286]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

saml

### Describe the bug

Some external IDP's that have a onboarding flow via there own redirect link to our saml onboarding, get an invalid destination , because they use an uppercase alias of the idp.. Everything goes ok until the Destinationvalidator validaties with equals...

Fixed it temp. by overriding the destinationValidator and adding an toLowerCase on expected and actual destionations.

### Version

26.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

to my opinion onboarding shouldn't be failing because of wrong alias idp casing ?

### Actual behavior

saml onboarding of new employees succeed.

### How to Reproduce?

do an idp onboarding (i.e.e Azure ?), and make sure the post to the /broker /endpoint has an uppercase alias name while the idp is configured lowercase.

Also made a Quarkus test to test the validator. WIthout my fix the second test fails

@QuarkusTest
public class DestinationValidatorTest {

    private DestinationValidator destinationValidator;

    @BeforeEach
    public void setup() {
        destinationValidator = DestinationValidator.forProtocolMap(new String[]{"http=8080", "https=443"});
    }

    @Test
    public void testExpectedActualCorrectBothLowercaseIdp() {

        String expectedDestination = "https://login.fdmg.nl/auth/realms/fd/broker/xxxx/endpoint";
        String actualDestination = "https://login.fdmg.nl/auth/realms/fd/broker/xxxx/endpoint";

        boolean result = destinationValidator.validate(expectedDestination, actualDestination);

        assertTrue(result);
    }

    @Test
    public void testExpectedActualCorrectButCaseSensitive() {

        String expectedDestination = "https://login.fdmg.nl/auth/realms/fd/broker/xxxx/endpoint";
        String actualDestination = "https://login.fdmg.nl/auth/realms/fd/broker/XXXX/endpoint";

        boolean result = destinationValidator.validate(expectedDestination, actualDestination);

        assertTrue(result);
    }
    }

**Comment by jonkoops** [comment]
@rvanderboom sounds like you have a pretty good idea of the problem. Would you be willing to submit a patch to fix this issue?

**Comment by rvanderboom** [comment]
yes, but never done, what do i have to deliver, is there some manual how to do it , and what to deliver ?

I made a quick fix only via, making both destinations lowercase (see below), but the base method (that receives URI), also has an else case if the pure string compare does not match , and then builds up an updateRequest with port nr etc, but what i see that alternative ALWAYS returns false since it is building a url like :  https://a.b.c:443/... 
but the :443 makes it never give succeeded because it doesn do this with actual one that passes just https://a.b.c/...

But my oppinion is it can be deviced in my fix to at least fix the saml validations, don't know if there is code that calls the methods with URI directly and if somewhere this causes another bug...



   public boolean validate(String expectedDestination, String actualDestination) {
        try {
            return this.validate(expectedDestination == null ? null : URI.create(expectedDestination.toLowerCase()),
                    actualDestination  == null ? null : actualDestination.toLowerCase());
        } catch (IllegalArgumentException var4) {
            return false;
        }
    }

**Comment by jonkoops** [comment]
If you'd like to contribute a fix I would recommend starting with the [contribution guide](https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md), it should have all the information you need. Once you have a PR up we can discuss implementation details and whether they are sufficient.

Let me know if you are interested and I will assign the issue to you.

**Comment by rvanderboom** [comment]
well also here to much prio on other things since a simple fix like above (case insensitive check) fixes it...
for us not issue to override some keycloak code since we more often already do on saml class since we have added more logging then the additional trace/debug/info logs, we have 120 idp's with different setups so we need more logging anyway..

And for an approved fix i think also the last part of the Validator (where the port number is involved) is not full prove correct also, but for us the quick fix works.

So can only pickup if gotten time, sorry..
